resteasy-client/src/main/java/org/jboss/resteasy/client/jaxrs/internal/ClientInvocation.java

```diff
@@ -589,9 +545,92 @@ public T extractResult(ClientResponse response)
                return (T) response;
             return ClientInvocation.extractResult(responseType, response, null);
          }
+      };
+
+      ClientHttpEngine httpEngine = client.httpEngine();
+      if (httpEngine instanceof AsyncClientHttpEngine)
+      {
+         return createAsyncSubmitFunction(true, callback, extractor).apply((AsyncClientHttpEngine) httpEngine);
+      }
+      else
+      {
+         // never buffered, but always blocks in a thread
+         final Function<ExecutorService, Future<T>> executorSubmitFn = executorService -> executorSubmit(executorService, callback, extractor);
+         return executorSubmitFn.apply(client.asyncInvocationExecutor());
+      }
+   }
+
+   private <T> Function<AsyncClientHttpEngine, Future<T>> createAsyncSubmitFunction(final boolean buffered,
+         final InvocationCallback<T> callback, final AsyncClientHttpEngine.ResultExtractor<T> extractor)
+   {
+      return asyncClientHttpEngine -> asyncSubmit(
+            ext -> (callback != null
+                  ? asyncClientHttpEngine.submit(this, buffered, callback, ext)
+                  : asyncClientHttpEngine.submit(this, buffered, ext, client.asyncInvocationExecutor())),

```

Comment:
For the scenarios where a `Future` would be returned but the `callback` is not passed, it would cause the `CompletableFuture` path to be taken.  It may not be the desirable thing to do as the default implementation in `AsyncHttpClientEngine` for `CompletableFuture` blocks on `Future`.  So, for the engines that do not override `CompletableFuture` version of `submit`, the execution will now be blocking.

Reply:
OK, please bear with me and help me understand better.
Are you basically saying that we should really keep separate `submit` / `submitCF` methods in `ClientInvocation`? (in the refactoring commit I've removed the `submitCF` ones except for the one that accept a callback as parameter).
We haven't overridden the `CompletableFuture` versions of submit in `AsyncHttpClientEngine` anywhere. I was actually a bit puzzled about that earlier, with the changes for this jira we're removing the blocking get in the `CompletionStageRxInvokerImpl`, but adding it in the default methods in `AsyncHttpClientEngine`: can you clarify how is that resulting in an improvement? Thanks

Reply:
> ... how is that resulting in an improvement?

It allows `AsyncClientHttpEngine` implementation to override the following and not block:

```java
default <T> CompletableFuture<T> submit(ClientInvocation request, boolean buffered, ResultExtractor<T> extractor)
```

For instance, I am working on a Reactor Netty based `ReactorNettyClientHttpEngine` which will implement the `default` method in `AsyncClientHttpEngine` and it won't block in `default <T> CompletableFuture<T> submit` method (I will send a PR in a few weeks).

You are right, with the current `ApacheHttpAsyncClient4Engine`, it does not make a difference.  But, in later versions, `ApacheHttpAsyncClient4Engine` can implement the `default <T> CompletableFuture<T> submit` method, and not block in it.  It should not be much work and can be achieved as part of this effort.  However, there is the `JettClientEngine` as well.  Apart from these, there may be some other custom `AsyncClientHttpEngine` that RestEasy users might have implemented on their own.  I would not want to create a regression for them.

> Are you basically saying that we should really keep separate `submit` / `submitCF` methods in `ClientInvocation`? 

It is not ideal; however, we may have to wait until we can get to a state that we do not need to provide `default` implementations in `AsyncClientHttpEngine`.  The `callback` itself cannot be the factor to decide which `submit` method in `AsyncClientHttpEngine` to call.

Reply:
OK, great, thanks a lot for the clarification. I'll revert the 2 refactoring commits later tonight and see if I can cleanup the code a bit without merging the two execution paths (`Future` and `CompletableFuture` ones). Will tag you for an additional review if you don't mind :-)

I'm really looking forward to see review your future PR(s) for new `AsyncClientHttpEngine` implementations.

Reply:
Sure!  Looking forward to the update!